src/antsibull_core/collection_meta.py

```diff
@@ -0,0 +1,161 @@
+                >= removal.major_version
+            ):
+                self.errors.append(
+                    f"{prefix} redirect_replacement_major_version: Redirect removal major version"
+                    f" {removal.redirect_replacement_major_version} must be smaller than"
+                    f" the removal major version {removal.major_version}"
+                )
+
+        if removal.reason == "renamed" and removal.new_name == collection:
+            self.errors.append(f"{prefix} new_name: Must not be the collection's name")
+
+    def _validate_collection(
+        self, collection: str, meta: CollectionMetadata, prefix: str
+    ) -> None:
+        if meta.repository is None:
+            self.errors.append(f"{prefix} repository: Required field not provided")
+
+        if meta.removal:
+            self._validate_removal(collection, meta.removal, f"{prefix} removal ->")
+
+    def validate(self, data: CollectionsMetadata) -> None:
+        # Check order
+        sorted_list = sorted(data.collections)
+        raw_list = list(data.collections)
+        if raw_list != sorted_list:
+            for raw_entry, sorted_entry in zip(raw_list, sorted_list):
+                if raw_entry != sorted_entry:
+                    self.errors.append(
+                        "The collection list must be sorted; "
+                        f"{sorted_entry!r} must come before {raw_entry}"
+                    )
+                    break
+
+        # Validate collection data
+        remaining_collections = set(self.all_collections)
+        for collection, meta in data.collections.items():
+            if collection not in remaining_collections:
+                self.errors.append(
+                    f"collections -> {collection}: Collection not in ansible.in"
+                )
+            else:
+                remaining_collections.remove(collection)
+            self._validate_collection(
+                collection, meta, f"collections -> {collection} ->"
+            )
+
+        # Complain about remaining collections
+        for collection in sorted(remaining_collections):
+            self.errors.append(f"collections: No metadata present for {collection}")
+
+
+def lint_collection_meta(
+    *, collection_meta_path: StrPath, major_release: int, all_collections: list[str]
+) -> list[str]:
+    """Lint collection-meta.yaml."""
+    validator = _Validator(
+        all_collections=all_collections,
+        major_release=major_release,
+    )
+
+    for cls in (
+        # NOTE: The order is important here! The most deeply nested classes must come first,
+        #       otherwise extra=forbid might not be used for something deeper in the hierarchy.
+        RemovalInformation,
+        CollectionMetadata,
+        CollectionsMetadata,
+    ):
+        cls.model_config["extra"] = "forbid"
+        cls.model_rebuild(force=True)
+
+    if not os.path.exists(collection_meta_path):
+        return [f"Cannot find {collection_meta_path}"]
```

Comment:
Noticing this now: it would be more efficient if this was moved to the top of the function.

Reply:
Originally the path was only composed at this point, so it maked sense to have that test there. Since it's now passed in, it definitely makes sense!

Reply:
de87a96da267a1d2e4cc9a074a6596df30627e49